http-clients/src/main/java/org/triplea/http/client/GenericWebSocketClient.java

```diff
@@ -0,0 +1,147 @@
+ * Furthermore, this class provides a listener API to be notified of incoming message events, a send
+ * API, and it automatically converts incoming and outgoing messages to JSON string.
+ *
+ * @param <IncomingT> Message type we expect to receive from the server.
+ * @param <OutgoingT> Message type we send to the server.
+ */
+// TODO: test-me
+@Log
+public class GenericWebSocketClient<IncomingT, OutgoingT> {
+  private static final Gson gson = new Gson();
+  private static final int CONNECT_TIMEOUT_MILLIS = 5000;
+
+  private final WebSocketClient client;
+  private final Collection<Consumer<IncomingT>> messageListeners = new HashSet<>();
+
+  private final ExecutorService threadPool;
+  private boolean connectionIsOpen;
+  private boolean closed = false;
+
+  public GenericWebSocketClient(
+      final Class<IncomingT> incomingMessageType,
+      final Class<OutgoingT> outgoingMessageType,
+      final URI lobbyUri,
+      final Consumer<String> connectionLostListener) {
+    threadPool = Executors.newFixedThreadPool(4);
+
+    client =
+        new WebSocketClient(lobbyUri) {
+          @Override
+          public void onOpen(final ServerHandshake serverHandshake) {}
+
+          @Override
+          public void onMessage(final String message) {
+            notifyListeners(message, incomingMessageType);
+          }
+
+          @Override
+          public void onClose(final int code, final String reason, final boolean remote) {
+            connectionLostListener.accept(reason);
+          }
+
+          @Override
+          public void onError(final Exception ex) {
+            connectionLostListener.accept(ex.getMessage());
+          }
+        };
+    client.connect();
+  }
+
+  private void notifyListeners(final String message, final Class<IncomingT> messageType) {
+    final IncomingT converted = gson.fromJson(message, messageType);
+    messageListeners.forEach(listener -> listener.accept(converted));
+  }
+
+  /**
+   * Non-blocking send of a message to the server. Implementation note: data is sent as a JSON
+   * string, this method handles conversion of the parameter object to JSON.
+   *
+   * @param message The data object to send to the server.
+   */
+  public void send(final OutgoingT message) {
+    final String json = gson.toJson(message);
+    final Future<?> result = threadPool.submit(() -> sendMessage(json));
+
+    // wait for result on a new thread so we do not block the current one
+    // this is only done so we can report any errors that have occurred.
+    new Thread(
+            () -> {
+              try {
+                result.get();
+              } catch (final InterruptedException e) {
+                Thread.currentThread().interrupt();
+              } catch (final ExecutionException e) {
+                log.log(Level.SEVERE, "Failed sending message: " + e.getMessage(), e);
+              }
+            })
+        .start();
```

Comment:
java-extras should be available everywhere.

That utility is good to know about, I'm not sure if 'Void' is going to be the right type. Regardless, we may likely still need a new thread; error-prone is stating that the 'result' needs to be checked, which will need to be done in a try/catch block and to avoid blocking in a new thread.

Reply:
`logWhenComplete` looks to be the thing. I did not find a `CompletableFuture.runAsync(...`


```
    CompletableFutureUtils.logExceptionWhenComplete(
        CompletableFuture.runAsync(() -> sendMessage(json), threadPool),
        "Failed to send message to server");
```

Reply:
What do you mean by "did not find"?
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/CompletableFuture.html#runAsync(java.lang.Runnable,java.util.concurrent.Executor)

Reply:
You still get the error prone error as 'sendMessage' returns future that is ignored.
`() -> sendMessage(json)`

The API you noted takes a runnable rather than Callable, the return value is not actually used.